packages/ai-jsx/src/batteries/docs.tsx

```diff
@@ -351,166 +380,213 @@ export interface EmbeddedChunk<ChunkMetadata extends Jsonifiable = Jsonifiable>
-  chunk: EmbeddedChunk<ChunkMetadata>;
+  /** The Chunk result. */
+  chunk: Chunk<ChunkMetadata>;
+  /** The Chunk's raw score for the given query. */
   score: number;
+
+  /** The id of the vector encoding this chunk (if supported). */
+  id?: string;
 }
 
 /**
  * A corpus is a collection of documents that can be provided to an LLM to provide grounding for
- * responses, along with the machinery to load and index those documents.
+ * responses.
  */
 export interface Corpus<ChunkMetadata extends Jsonifiable = Jsonifiable> {
-  /**
-   * Begins loading documents into the corpus using a Loader, Chunker, and Embedding.
-   *
-   * Note: This returns a promise because some implementations may make an external request
-   * to begin loading. However, the promise will typically resolve prior to the corpus being fully
-   * loaded. In such cases, use `getStats` to check progress.
-   */
-  startLoading: () => Promise<Corpus.Stats>;
-
-  /**
-   * Reveals statistics about this corpus, including its LoadingState.
-   *
-   * Note: This returns a promise because some implementations may make an external request.
-   */
-  getStats: () => Promise<Corpus.Stats>;
-
-  /**
-   * Finds document chunks that are semantically similar to the provided query.
-   */
+  /** Finds document chunks that are semantically similar to the provided query. */
   search: (
     query: string,
     params?: { limit?: number; score_threshold?: number }
   ) => Promise<ScoredChunk<ChunkMetadata>[]>;
 }
 
-export namespace Corpus {
-  /** An indication of what steps are necessary to make this corpus ready for serving. */
-  export enum LoadingState {
-    /** Documents have not been loaded. Call `startLoading` to begin loading them. */
-    NOT_STARTED = 'NOT_STARTED',
-    /** Documents are currently being loaded. */
-    IN_PROGRESS = 'IN_PROGRESS',
-    /** Documents were successfully loaded. The corpus is ready for serving. */
-    COMPLETED = 'COMPLETED',
-    /** Documents failed to load. See details for your implementation to try to resolve. */
-    FAILED = 'FAILED',
-  }
+/** An indication of what steps are necessary to make this corpus ready for serving. */
+export enum CorpusLoadingState {
+  /** Documents have not been loaded. Call `startLoading` to begin loading them. */
+  NOT_STARTED = 'NOT_STARTED',
+  /** Documents are currently being loaded. */
+  IN_PROGRESS = 'IN_PROGRESS',
+  /** Documents were successfully loaded. The corpus is ready for serving. */
+  COMPLETED = 'COMPLETED',
+  /** Documents failed to load. See details for your implementation to try to resolve. */
+  FAILED = 'FAILED',
+}
+
+/** Statistics about a corpus, including its LoadingState. */
+export interface CorpusStats {
+  loadingState: CorpusLoadingState;
+  completedPartitions: number;
+  activePartitions: number;
+  numDocuments: number;
+  numChunks: number;
+  /** If loadingState == FAILED, the error that caused it. */
+  loadingError?: Error;
```

Comment:
We can use TS to encode that `loadingError` will only appear when `loadingState === failed`:

https://www.typescriptlang.org/play?#code/PQKhAIEEDtwS2gEzgYwIYBc4HtbYGbgDuAFpuAM4YCmADheGgE7XjTUrUUXMCe4GbOAC2aANasMJOAxTYmtAK4MWaRP3zzK1JgDcEAcwB04EMABQ1AB615GcNWiLh4ADLY1hgMoZMrAN7m4OCgEAAi2CjOjhgMZLqs0Nj2AEbUjuAANh6I1IgmAMJomZngAAZUzBjuntAGZQJCaQYIWTmGAiTUwiZmQeAAcgDyACoA+l4jkABKIwCiYeAAvOAA5MPjkzPzYasANP2h4BFRwjEMzKxRTCzQGJn8aR3Zanm9FsEAkgNjAArTQwA4tM5l4vMs1t8-gDgaCvPtDmBjpFoncGEQdKwKIoUJxuPhFCV+C9cvlwCMuuA5AplPAVNQ1BotBQdPo6u9+gUhgBZX6uOY7CGrLm8-k7BHBI4nVGxcD4NBwTJ5RptNQmLzpcC5XyKhiaJjgXjYRQGuDCWhKs53TA4WCCARMfj2lgUbCZBIc4IAMUgn35ixWqx9foWCIAvuZzEcfDaqKgLiljfY0FT5EoKHt4NAUJlFMg6vBZTV8wYYzQOQgaEx5ZxwAAhNAsssMQLBEneXw0ABcbnadTL1AA3OYI+ZKzoa6xmwB1OBSOY3LTWGhIBgNpudlv9dv9zvUHvFjt+IzB-3Dtt9gwLpjyHvX+TD0fj6toWszuckJP3g3LxyINeNtQzbgK2qolgOd5WDmebUAAPIeu5+JmCGlnuJ6+v6AB8I6RhgvC0FOm4Qu+86LgaAA+4AkZ+igYN+kbAMA5LSHqCqZAwaToMorAYqsLAiDIFDPJe9FyNAVCUD2wErKBO6oX4B6XgO6EhmEOFRkxFIyHKbEcRwaDccQrDxKwcnfuAKS0ZQNGZIg0CrBg5hiRJFAAExSURMnbkpe6KbU8nlhsExTLMCwHBe-nfj27BEOA34ABSrPKYirAAlOpjHMdpRDyGIFBObgLkAMweZgDBeRF4G+b2-nKaeYXeZFZHRdQsUJUl4hpThQA

Reply:
This is super cool. Thanks for sharing!

Unfortunately this seems to keep me from updating individual fields on the stats object, which makes it a bit clunky to use unless there's a trick I'm missing.